controllers/storagecluster/storageclient.go

```diff
@@ -0,0 +1,145 @@
+
+	storageClientCRD := &extv1.CustomResourceDefinition{}
+	storageClientCRD.Name = storageClientCRDName
+	err := r.Get(r.ctx, client.ObjectKeyFromObject(storageClientCRD), storageClientCRD)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			r.Log.Info("Waiting for StorageClient CRD to be present")
+			// if we are running in provider mode, deployment of storageclient is mandatory and so poll for crd
+			return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+		}
+		r.Log.Error(err, "Failed to verify existence of StorageClient CRD")
+		return reconcile.Result{}, err
+	}
+
+	storageClient := &ocsclientv1a1.StorageClient{}
+	storageClient.Name = storageClientName
+	storageClient.Namespace = r.OperatorNamespace + "-client"
+	_, err = cutil.CreateOrUpdate(r.ctx, r.Client, storageClient, func() error {
+		if storageClient.Spec.OnboardingTicket == "" {
+			token, err := getOnboardingToken()
+			if err != nil {
+				return fmt.Errorf("unable to get onboarding token: %v", err)
+			}
+			storageClient.Spec.OnboardingTicket = token
+		}
+		storageClient.Spec.StorageProviderEndpoint = storagecluster.Status.StorageProviderEndpoint
+		return nil
+	})
+	if err != nil {
+		r.Log.Error(err, "Failed to create local StorageClient CR")
+		return reconcile.Result{}, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (s *storageClient) deleteStorageClient(r *StorageClusterReconciler) (reconcile.Result, error) {
+	storageClient := &ocsclientv1a1.StorageClient{}
+	storageClient.Name = storageClientName
+	storageClient.Namespace = r.OperatorNamespace + "-client"
+	if err := r.Delete(r.ctx, storageClient); err != nil && !kerrors.IsNotFound(err) {
+		r.Log.Error(err, "Failed to initiate deletion of local StorageClient CR")
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// getOnboardingToken makes an api call to ux-backend with service account token and gets onboarding token
+func getOnboardingToken() (string, error) {
+	saToken, err := os.ReadFile(saTokenPath)
+	if err != nil {
+		return "", err
+	}
+
+	tokenURL := fmt.Sprintf("https://%s:%d/%s", ocsinit.UxBackendSvcName, ocsinit.UxBackendSvcPort, onboardingTokenURLPath)
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		},
+		Timeout: 30 * time.Second,
+	}
+	tokenRequest, err := http.NewRequest(http.MethodPost, tokenURL, nil)
+	if err != nil {
+		return "", err
+	}
+	tokenRequest.Header.Set("Authorization", fmt.Sprintf("Bearer %s", string(saToken)))
+
+	tokenResponse, err := client.Do(tokenRequest)
+	if err != nil {
+		return "", err
+	}
+	defer tokenResponse.Body.Close()
+	token, err := io.ReadAll(tokenResponse.Body)
+	if err != nil {
+		return "", err
+	}
+	if tokenResponse.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("api call to token endpoint resulted in status code: %d", tokenResponse.StatusCode)
+	}
+
+	return string(token), nil
+}
```

Comment:
Should we really have the ocs-operator call into the ux-backend? Can we refactor the token generation code as a util func where we call it both from here and from the ux-backend endpoint?

Reply:
I think that might defeat the purpose of isolation that ocs-operator shouldn't read the private key 🤔

Reply:
1. The bigger risk is ocs-operator writing the private key, not reading it.
2. There was no use case for reading it before, now there is.
3. By relegating to the ux server you solved nothing. If someone has access to ocs-operator they can issue calls into the ux-backend to get onboarding key, it is just another level of indirection for the read.

Reply:
Ack, it's fine then. BTW, we are mandating user to not manage private key, isn't it?